### DISCUSSION TITLE
How can I improve localization in the Avalonia app?

### DISCUSSION THREAD

**Discussion by SignalMQ** [AvaloniaUI/Avalonia#13971] [Q&A]
**Hi everybody!**
Please tell me how you can implement localization of the application in such a way that when you call the button with languages, you can change the translation in the window without reloading the application.

**Which ways have I tried before?**
- I used the nuget library from `NP.Visuals`. But it has recently ceased to be supported by the author, since the release of the latest version of `Avalonia`.
- I tried to use `.resx` resources. Since the resources use `CultureInfo`, I had to restart the application for the translation to apply to the entire environment.

**Comment by cemalgulbeyaz** [comment]
I have an idea about it. Maybe this is not the best way for that but i had to decide use like this way:

LocalStringStructure.cs this to marking structure of  your language strings. Inner General class to categorize your strings. It can also more Inner too. 
```
public static partial class LocalStringStructure
{
    public partial class General : ObservableObject, IDeepCopyable
    {
        [ObservableProperty] private string title;
    }
}
```
LocalStrings .cs this your base language it has static instance to reach from xaml binding. And there is a task load json file from your source folder by given language code.
```
    public partial class LocalStrings : ObservableObject, IDeepCopyable
    {
        public static LocalStrings Instance = new();

        [ObservableProperty] private string languageCode = "default";
        [ObservableProperty] private string name;
        [ObservableProperty] private string version;
        [ObservableProperty] private LocalStringStructure.General general;

        public LocalStrings()
        {
            LanguageCode = "en";

            Version = "1.0";

            Name = "English";

            General = new LocalStringStructure.General { 
                Title = "Test Title",
            };
        }


        public static async Task Load(string languageCode)
        {
            LocalStrings? localStrings = Instance;

            if (languageCode != null)
            {
                string baseDirectory = AppDomain.CurrentDomain.BaseDirectory;
                string jsonFolderPath = Path.Combine(baseDirectory, nameof(Languages));
                if (!Path.Exists(jsonFolderPath))
                    throw new Exception("Languages folder couldn't found in application folder.");

                var jsonFileName = $"{languageCode}.json";
                var jonFilePath = Path.Combine(jsonFolderPath, jsonFileName);

                if (Path.Exists(jonFilePath))
                {
                    string jsonContent = await File.ReadAllTextAsync(jonFilePath);
                    localStrings = JsonConvert.DeserializeObject<LocalStrings>(jsonContent);
                }
            }
            else
            {
                localStrings = new();
            }

            if (localStrings != null)
                localStrings.DeepCopyTo(Instance);
        }
    }
```

Example binding
`Text="{Binding Source={x:Static LocalStrings.Instance}, Path=General.Title}"`

IDeepCopyable.cs it to marking extension method and IStartUpLanguage.cs creating json files in AppBuilder from project classes
```
public interface IDeepCopyable
{
}

public interface IStartUpLanguage
{
}
```
Extensions.cs contains extension methods, DeepCopyTo method for language changes in Load method. CreateLocalStringJsons method to create json files from classes. Yo need add this extension method end of the "AppBuilder.Configure<App>()" tail in Program.cs
```
public static class Extensions
{
    public static void DeepCopyTo<T>(this T source, T destination) where T : IDeepCopyable
    {
        if (source == null)
            throw new ArgumentNullException(nameof(source));

        if (destination == null)
            throw new ArgumentNullException(nameof(destination));

        PropertyInfo[] properties = typeof(T).GetProperties();

        foreach (var property in properties)
        {
            if (property.CanRead && property.CanWrite)
            {
                var value = property.GetValue(source);
                property.SetValue(destination, value);
            }
        }
    }
        public static AppBuilder CreateLocalStringJsons(this AppBuilder builder)
        {
            string baseDirectory = AppDomain.CurrentDomain.BaseDirectory;
            string jsonFolderPath = Path.Combine(baseDirectory, nameof(Languages));
            if (!Path.Exists(jsonFolderPath)) 
                Directory.CreateDirectory(jsonFolderPath);

            var interfaceType = typeof(IStartUpLanguage);
            var startUpLanguages = AppDomain.CurrentDomain.GetAssemblies()
                                                          .SelectMany(s => s.GetTypes())
                                                          .Where(p => interfaceType.IsAssignableFrom(p) && p.IsClass);

            foreach (Type startUpLanguage in startUpLanguages)
            {
                var instance = Activator.CreateInstance(startUpLanguage);
                var localStringsField = startUpLanguage.GetFields()
                                                       .FirstOrDefault(f => f.Name == nameof(LocalStrings));
                var localStrings = localStringsField?.GetValue(instance) as LocalStrings;

                if (localStrings == null)
                    continue;

                var jsonFileName = $"{localStrings.LanguageCode}.json";
                var jonFilePath = Path.Combine(jsonFolderPath, jsonFileName);

                if (Path.Exists(jonFilePath))
                    continue;

                using (StreamWriter file = File.CreateText(jonFilePath))
                {
                    JsonSerializer serializer = new JsonSerializer();
                    serializer.Formatting = Formatting.Indented;
                    serializer.Serialize(file, localStrings);
                }
            }

            return builder;
        }
}
```
LocalStringsTr.cs and example extra language json file will be created by this class. You can also add directly json file too with correct structure.
```
    public class LocalStringsTr : IStartUpLanguage
    {
        public static readonly LocalStrings LocalStrings = new LocalStrings
        { 
            LanguageCode = "tr",

            Version = "1.0",

            Name = "Türkçe",

            General = new LocalStringStructure.General
            {
                Title = "Test Başlık",
            }
        };
    }
```

**Reply by SignalMQ** [reply]
@cemalgulbeyaz thanks for the answer.

**Comment by SignalMQ** [comment]
During this time, I rethought and redid the work with `.resx ` files. It turned out quite well in my opinion, of course, this functionality may lack a lot of things, but this is enough for me.
1) I've added `.resx` files with three localization options:
![image](https://github.com/AvaloniaUI/Avalonia/assets/77734001/a4cfa631-dbda-4f4e-9767-2c6eab9d742c)
2) My `XAML` markup looks like this:
```
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:LocalizationSampleApp.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="LocalizationSampleApp.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel"
        Icon="/Assets/avalonia-logo.ico"
        Title="LocalizationSampleApp">

    <Design.DataContext>
        <!-- This only sets the DataContext for the previewer in an IDE,
             to set the actual DataContext for runtime, set the DataContext property in code (look at App.axaml.cs) -->
        <vm:MainWindowViewModel/>
    </Design.DataContext>

    <Grid ColumnDefinitions="*,Auto,Auto,Auto,*" RowDefinitions="*,Auto,Auto,*">
        <TextBlock Grid.Column="1" Grid.ColumnSpan="3" Grid.Row="1" Text="{Binding Greeting}" HorizontalAlignment="Center" VerticalAlignment="Center"/>
        <Button Grid.Column="1" Grid.Row="2" HorizontalAlignment="Center" Content="EN" Command="{Binding SetLanguageEn}"/>
        <Button Grid.Column="2" Grid.Row="2" HorizontalAlignment="Center" Content="RU" Command="{Binding SetLanguageRu}"/>
        <Button Grid.Column="3" Grid.Row="2" HorizontalAlignment="Center" Content="UZ" Command="{Binding SetLanguageUz}"/>
    </Grid>
</Window>
```
3) Next, my MVVM implementation for this window looks like this:
```
using System.Globalization;
using System.Reactive;
using ReactiveUI;

namespace LocalizationSampleApp.ViewModels;

public class MainWindowViewModel : ViewModelBase
{
#pragma warning disable CA1822 // Mark members as static
    private string _greeting = Resources.Resources.Greeting;
    public string Greeting
    {
        get => _greeting;
        set => this.RaiseAndSetIfChanged(ref _greeting, value);
    }
#pragma warning restore CA1822 // Mark members as static

    public ReactiveCommand<Unit, Unit> SetLanguageEn =>
        ReactiveCommand.Create(() => SetLanguage(new CultureInfo("en-US")));
    public ReactiveCommand<Unit, Unit> SetLanguageRu =>
        ReactiveCommand.Create(() => SetLanguage(new CultureInfo("ru-RU")));
    public ReactiveCommand<Unit, Unit> SetLanguageUz =>
        ReactiveCommand.Create(() => SetLanguage(new CultureInfo("uz-Latn")));

    void SetLanguage(CultureInfo? cultureInfo)
    {
        Resources.Resources.Culture = cultureInfo;
        Greeting = Resources.Resources.Greeting;
    }
}
```

It so happened that the resources, as it turned out, use their own internal `CultureInfo`, which allowed me to change the language of the resources themselves without changing the culture of the environment. But I'm not sure whether the translation will extend to other calls to `Resources.Resources.<some_word>` from other windows.

**Reply by cemalgulbeyaz** [reply]
I was not interested in .resx only because it does not allow adding and editing languages during development.
In the code I shared, I may have made a slight misnomer for the interface, which I called IStartUpLanguage. IBuiltInLanguage would be better.
Since the strings are added from external json files, and the structure of json files is simple and allows editing by those with little knowledge, it made more sense to me.
When the Load method is called with a language code, if there is a json file named with the language code in the source folder (Languages), the strings are read as LocalStrings and assigned to the target property (Instance) in the DeepCopyTo extension method. After this assignment, the values of all XAML objects bound with Binding are changed.
By the way, in your code, the assignment "Greeting = Resources.Resources.Resources.Greeting;" in the SetLanguage method will force you as the number of strings added increases. Here, a structure similar to what I did in DeepCopyTo will make your job easier. In my first attempts, I did manual mapping like you, and then I made this change.

**Reply by memorydream** [reply]
I remember have a simple solution for use resx and binding , should can get reference from [SteamTools](https://github.com/BeyondDimension/SteamTools) , you can try find it

**Comment by timunie** [comment]
For smaller Apps you can also use DynamicResources. Check out tims.avalonia.samples for reference

**Comment by timunie** [comment]
Also https://github.com/AvaloniaUI/Avalonia/discussions/13448

**Comment by JaggerJo** [comment]
We've had the same issue with multiple apps. At some point we decided to just roll our own and add features as needed.

- Language change at runtime
- Strongly typed translation keys (no string keys)
- Support for multiple translation files (per feature, ..)

https://github.com/Voyonic-Systems/Echoes

**Comment by David583** [comment]
Any improvements with localization?

I really like this framework, and want to do things nicely, but I feel like I have already put too much time in this localization problem than I wanted to.

I have managed to change the language on runtime by playing with the resources, now I only have one problem:

How can I insert variables into my translations? Do you have some simple way like i18next, something like this:

`You have {{  itemCount }} items in your cart`

**Reply by JaggerJo** [reply]
Take a look at Echoes

https://github.com/Voyonic-Systems/Echoes

**Reply by David583** [reply]
Already did that. Although, it solves all of my problems, I personally doesn't like the way it is implemented. First, the translation needs to be a separate project. That's uh... strange and inconvenient, but I am sure you had the reasons to implement it the way it is. Second, I am working with net 9.0, I don't want to downgrade. Third, you are working with toml files, that's an another strange pick for me. I would prefer JSON, or just stay at AXML. Fourth, I find it strange the way you implemented the things in the View side. I took a look at the example project, and found this: 

```
<DockPanel>
          <TextBlock
              Text="{echoes:Translate {x:Static translations:Strings.greeting}}"/>

          <TextBox
              Text="{Binding Name}"/>

</DockPanel>
```

For me, this means that the way you are inserting variables into translations, is that they need to be in a same container element, but in separate text elements? It's kinda inconvenient, couldn't you do something like this:

`<TextBlock
                Text="{echoes:Translate {x:Static translations:Strings.greeting, variables:Binding Name}}"/>`

Again, I am sure that you had all the reasons to implement things in the way it is. All I am saying, that I am looking for a simple, built in way to handle translations, that can handle runtime culture change and variable insertion.

**Reply by JaggerJo** [reply]
> Already did that. Although, it solves all of my problems, I personally doesn't like the way it is implemented. 

> First, the translation needs to be a separate project. That's uh... strange and inconvenient, but I am sure you had the reasons to implement it the way it is. 

That's because a source generator is used to generate static classes for the translation keys. Avalonia uses a Source Generator for XAML. Because generated code can't consume generated code it needs to be in a separate project. Otherwise you can't have compile time checking for translation keys.

> Second, I am working with net 9.0, I don't want to downgrade. 

You don't need to. You can consume .net 8 libraries from .net 9 without any issues.

> Third, you are working with toml files, that's an another strange pick for me. I would prefer JSON, or just stay at AXML. 

JSON would be worth considering but more verbose. For AXAML I'm not sure how you'd use that to store translation values.

> Fourth, I find it strange the way you implemented the things in the View side. I took a look at the example project, and found this:
> 
> ```
> <DockPanel>
>           <TextBlock
>               Text="{echoes:Translate {x:Static translations:Strings.greeting}}"/>
> 
>           <TextBox
>               Text="{Binding Name}"/>
> 
> </DockPanel>
> ```

That's just one way to do it, others are perfectly valid. 

> For me, this means that the way you are inserting variables into translations, is that they need to be in a same container element, but in separate text elements? It's kinda inconvenient, couldn't you do something like this:
> 
> `<TextBlock Text="{echoes:Translate {x:Static translations:Strings.greeting, variables:Binding Name}}"/>`

Not exactly, but something similar would be possible.

> Again, I am sure that you had all the reasons to implement things in the way it is. All I am saying, that I am looking for a simple, built in way to handle translations, that can handle runtime culture change and variable insertion.

Echoes was created because there was nothing I wanted to use. ResX makes it hard to change the language without reloading views.

**Reply by kyuranger** [reply]
I used markup extension and frozendictionary to achieve this. It works very well.

**Reply by Adam-Wilkinson** [reply]
I just started up with Echoes, I like this approach to localization in bindings. My solution to the "You have {{  itemCount }} items in your cart" problem was a small custom MultiValueConverter:

```
public class FormatStringConverter : IMultiValueConverter
{
    public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
    {
        if (values[0] is not string formatString)
        {
            return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
        }

        return string.Format(culture, formatString, values.Skip(1));
    }
}
```
with usage
```
<MultiBinding Converter="{StaticResource FormatStringConverter}">
        <Binding Path="Value^" Source="{x:Static lang:Strings.Greeting}"/>
        <Binding Path="Name"/>
</MultiBinding>
```
given toml
```
[translations]
Greeting = "Hello, {0}. Welcome!"
```

This gives the best of both worlds. We get an actual string out of it for use in a TextBlock, and with MultiBindings if either the translation string or the name are updated, so is the text. 

Using the Runs feature of TextBlock is also possible, but getting a single string out of the MultiBinding is much cleaner imo.

Just thought I'd leave this here because Echoes has saved me a lot of headache (other similar approaches such as tlumach are just as nice for the same reasons)

**Reply by JaggerJo** [reply]
> I just started up with Echoes, I like this approach to localization in bindings. My solution to the "You have {{  itemCount }} items in your cart" problem was a small custom MultiValueConverter:
> 
> 
> 
> ```
> 
> public class FormatStringConverter : IMultiValueConverter
> 
> {
> 
>     public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
> 
>     {
> 
>         if (values[0] is not string formatString)
> 
>         {
> 
>             return new BindingNotification(new InvalidCastException(), BindingErrorType.Error);
> 
>         }
> 
> 
> 
>         return string.Format(culture, formatString, values.Skip(1));
> 
>     }
> 
> }
> 
> ```
> 
> with usage
> 
> ```
> 
> <MultiBinding Converter="{StaticResource FormatStringConverter}">
> 
>         <Binding Path="Value^" Source="{x:Static lang:Strings.Greeting}"/>
> 
>         <Binding Path="Name"/>
> 
> </MultiBinding>
> 
> ```
> 
> given toml
> 
> ```
> 
> [translations]
> 
> Greeting = "Hello, {0}. Welcome!"
> 
> ```
> 
> 
> 
> This gives the best of both worlds. We get an actual string out of it for use in a TextBlock, and with MultiBindings if either the translation string or the name are updated, so is the text. 
> 
> 
> 
> Using the Runs feature of TextBlock is also possible, but getting a single string out of the MultiBinding is much cleaner imo.
> 
> 
> 
> Just thought I'd leave this here because Echoes has saved me a lot of headache (other similar approaches such as tlumach are just as nice for the same reasons)

Nice to hear!

**Comment by emayevski** [comment]
With people struggling to localize Avalonia UIs using resources, I have created and tested an open-source library that enables localization of various XAML-based applications by binding the controls to translation units. The library supports various translation file formats and enables one to switch languages on the fly without restarting the app. Maybe you will want to look at it. The [library is on Github](https://github.com/Allied-Bits-Ltd/tlumach-net) .
